LeNet1989 with Commentary

nalzok opened the pull request:
## Description

Jax/Flax implementation of LeCun 1989's CNN. Since you asked for a notebook, I assume you would like to optimize for readability. As a result, I based my work on Flax's official [Annotated MNIST](https://flax.readthedocs.io/en/latest/notebooks/annotated_mnist.html) notebook, which includes a lot of comments and is very comprehensive. I have also added some sanity checks to demonstrate the shape of the parameters are correct.

```python
assert state.params['H1w']['kernel'].shape == (5, 5, 1, 12)
assert state.params['H1b']['bias'].shape == (8, 8, 12)
assert state.params['H2s1w']['kernel'].shape == (5, 5, 8, 4)
assert state.params['H2b']['bias'].shape == (4, 4, 12)
assert state.params['H3']['kernel'].shape == (4*4*12, 30)
assert state.params['H3']['bias'].shape == (30,)
assert state.params['Out']['kernel'].shape == (30, 10)
assert state.params['Out']['bias'].shape == (10,)
```

I have not added the modern adjustments yet to avoid bloating the notebook, but I can include some if that's desirable. In fact, I have experimented with making all units in the convolutional layers share their bias in addition to the weights, and using minibatch instead of SGD. Both adjustments don't seem to have a huge impact, though.

### Colab link

<!-- Add colab link that redirects to your notebook -->
<!-- 

For example, if your notebook is : 

https://github.com/codeboy5/probml-notebooks/blob/add-densenet-jax/notebooks-d2l/densenet_jax.ipynb  

Then colab notebook link is:

https://colab.research.google.com/github/codeboy5/probml-notebooks/blob/add-densenet-jax/notebooks-d2l/densenet_jax.ipynb 

-->

https://colab.research.google.com/drive/1N5qvJfmJnHgmfSdw8frU5C9J7gS_NgKX?usp=sharing

### Issue 

<!-- Link the issue you are solving -->

<!-- If the issue is from this repo, include directly with issue number -->
<!-- For example: 

#12

-->

<!-- If the issue is from another repo, you can include it using the regular linking mechanism -->
<!-- For example:

[Issue title](Issue link) 

-->

[convert LeNet1989 from Torch to JAX](https://github.com/probml/pyprobml/issues/698)

## Checklist:

- [X] Performed a self-review of the code
- [X] Tested on Google Colab.

## Potential problems/Important remarks

Apart from commentary and coding styles, there are no substantial differences between my implementation and the one in https://github.com/probml/probml-notebooks/pull/54, despite we worked independently.

<!-- If you have any important remarks for the reviewers to look at closer,  you can write them in detail here.  -->

murphyk commented:
LGTM. However Please remove layer_norm_jax.ipynb from this PR.
And make the colab public so I can check it out.

nalzok commented:
I have removed the domain restriction. Please check again if you can access the notebook on Colab.

https://colab.research.google.com/drive/1N5qvJfmJnHgmfSdw8frU5C9J7gS_NgKX?usp=sharing

I have also deleted the file `layer_norm_jax.ipynb`, so that you can merge my branch now.

murphyk commented:
Yes, I can read it now. This looks like a great introduction to flax. Thanks!
